### PR TITLE
semantics for AMO instructions in DIY

### PR DESCRIPTION
**Pull request by hugookeeffe**
Implementing the semantics of AMO instructions into DIY so that the post condition can consider this when determining its values

### CODE REVIEW

File: gen/cycle.ml [hunk 1]
```diff
@@ -726,7 +726,6 @@ let set_same_loc st n0 =
             let ord = CoSt.get_co st Ord in
             let ctag = CoSt.get_co st CapaTag in
             let cseal = CoSt.get_co st CapaSeal in
-            fprintf stderr "test 20 \n" ;
```
Comment (relokin) [inline gen/cycle.ml]: I think we might want to squash this commit to the previous one.

File: gen/AArch64Arch_gen.ml [hunk 2]
```diff
@@ -597,6 +597,26 @@ let get_ie e = match e with
...
+let compute_rmw r old co =
+   begin match r with
+    | LdOp op ->
+      begin match op with
+        | A_ADD -> old + co
+        | A_SMAX -> if old > co then old else co
+        | A_SMIN -> if old < co then old else co
+        | A_EOR | A_SET | A_CLR -> co
+      end
+    | StOp op ->
```
Comment (relokin) [inline gen/AArch64Arch_gen.ml:610]: Can you handle both LdOp and StOp together? It looks like, we handle the same operators.

File: gen/AArch64Arch_gen.ml [hunk 3]
```diff
@@ -597,6 +597,26 @@ let get_ie e = match e with
...
 
+
+let compute_rmw r old co =
+   begin match r with
+    | LdOp op ->
+      begin match op with
+        | A_ADD -> old + co
+        | A_SMAX -> if old > co then old else co
+        | A_SMIN -> if old < co then old else co
+        | A_EOR | A_SET | A_CLR -> co
```
Comment (relokin) [inline gen/AArch64Arch_gen.ml:608]: I know this is wip, but if in the end we don't add support for bitwise ops yet, maybe we could add a warning.

File: gen/cycle.ml [hunk 4]
```diff
@@ -463,13 +463,20 @@ module CoSt = struct
...
 
-  let set_cell st e = match e.bank with
-    | Ord ->
-       let old = st.co_cell.(0) in
-       let cell = E.overwrite_value old e.atom e.v in
-       let co_cell = Array.copy st.co_cell in
-       co_cell.(0) <- cell ;
-       {e with cell=co_cell;},{ st with co_cell; }
+  let set_cell ?(rv = -1) st n =
+    let e = n.evt in  match e.bank with
```
Comment (relokin) [inline gen/cycle.ml:467]: Just to keep this diff to a minimum, maybe we could move this let further down where it's used. Also many of the lines below we're only changing their indentation.

File: gen/cycle.ml [hunk 5]
```diff
@@ -463,13 +463,20 @@ module CoSt = struct
...
+    | Ord -> begin
+      let old = st.co_cell.(0) in
+      let cell = E.overwrite_value old e.atom e.v in
+      let co_cell = Array.copy st.co_cell in
+      match n.edge.E.edge with
```
Comment (relokin) [inline gen/cycle.ml:472]: can we simplify this to something like
```
let cell2 =
      match n.edge.E.edge with
      | E.Rmw rmw -> E.compute_rmw rmw old cell
      | _ -> if rv = -1 then cell else rv in
co_cell.(0) <- cell2;
{e with cell=co_cell;},{ st with co_cell; }
```

File: gen/cycle.ml [hunk 6]
```diff
@@ -719,6 +726,7 @@ let set_same_loc st n0 =
             let ord = CoSt.get_co st Ord in
             let ctag = CoSt.get_co st CapaTag in
             let cseal = CoSt.get_co st CapaSeal in
+            fprintf stderr "test 20 \n" ;
```
Comment (relokin) [inline gen/cycle.ml]: you could remove this, or squash the next patch into this one.

File: gen/cycle.ml [hunk 7]
```diff
@@ -741,16 +749,16 @@ let set_same_loc st n0 =
...
                    n.evt <- e ;
                    do_set_write_val st pte_val ns
                 | Tag|CapaTag|CapaSeal ->
-                   let st = CoSt.next_co st bank in
-                   let v = CoSt.get_co st bank in
-                   n.evt <- { n.evt with v = v; } ;
-                   let e,st = CoSt.set_tcell st n.evt in
-                   n.evt <- e ;
-                   do_set_write_val st pte_val ns
+                  let st = CoSt.next_co st bank in
```
Comment (relokin) [inline gen/cycle.ml:755]: I don't think there is any meaningful change here, is there? It's only indentation.

File: gen/edge.ml [hunk 8]
```diff
@@ -31,6 +31,7 @@ module type S = sig
   type atom
   type rmw
 
+
```
Comment (relokin) [inline gen/edge.ml:34]: Unintended change?

File: gen/cycle.ml [hunk 9]
```diff
@@ -782,9 +790,30 @@ let set_same_loc st n0 =
...
             | Code _ ->
-               do_set_write_val st pte_val ns
+              do_set_write_val st pte_val ns
             end
-        | Some R | Some J |None -> do_set_write_val st pte_val ns
+        | Some R ->
+          begin match n.evt.loc with
+            | Data _ ->
+              let bank = n.evt.bank in
+              begin match bank with
```
Comment (relokin) [inline gen/cycle.ml:798]: just to avoid too much nesting you could also do something like
```
begin match bank, n.evt.rmw with
| Ord, true ->
....
```

**Review by relokin — COMMENTED** [review]
Thanks for the fix Hugo!

Just a couple of comments from a quick look at the code. I will also try to give it a go later today.

File: gen/AArch64Arch_gen.ml [hunk 10]
```diff
@@ -597,6 +597,26 @@ let get_ie e = match e with
...
+      end
+    | StOp op ->
+      begin match op with
+        | A_ADD -> old + co
+        | A_SMAX -> if old > co then old else co
+        | A_SMIN -> if old < co then old else co
+        | A_EOR | A_SET | A_CLR -> co
+      end
+    | LrSc | Swp | Cas  -> co
+  end
```
Comment (maranget) [inline gen/AArch64Arch_gen.ml:618]: This `begin/end` is not rquired. ON ca write `let compute_rmw r old co = match ...` directly.

File: gen/AArch64Arch_gen.ml (hunk 2 above)
Comment (maranget) [reply gen/AArch64Arch_gen.ml:610]: Agreed, you can use the or-pattern `LdOp op|Stop op -> ,,,` to share the identical clauses.

File: gen/cycle.ml [hunk 11]
```diff
@@ -708,7 +715,7 @@ let set_same_loc st n0 =
 
   let tr_value e v = E.tr_value e.atom v
 
-  let rec do_set_write_val st pte_val = function
+  let rec do_set_write_val ?(rv = -1) st pte_val = function
```
Comment (maranget) [inline gen/cycle.ml:718]: I do not understand the logic behind this optional argument `rv`. The state argument `st` has two components of interest, value `v` (in practice `v` is the value bound to `Ord`  in `st.map`) and component `co_cell`. Component `co_cell` (here, in the non-neon case `co_cell.(0)`) represents the contents of memory after the write is performed. As regards our extension component v still follows the sequence if integers, 0,1,2,... as before. The computation of `co_cell` changes: it becomes `set.co_cell.(0)` op `v` (op is the operation performed by the AMO operation and `v` is the operand of the instructio, available from `set_cell` as `n.evt.v`,

Hence, the function `set_cell` does not need the extra `rv` argument as the old value of memory is available as `st.cell.(0)` and the operand is available as `n.evt.v`.

Notice that further adaptation may be  needed for mixed-size mode: the calls to functions `tr_value` at lign 741/748 below and `E.overwrite_value` in `set_cell` are concern mixed-size mod and are trivial in non-mixed-size mode. But mixed-size is yet another story.

File: gen/cycle.ml [hunk 12]
```diff
@@ -463,16 +463,16 @@ module CoSt = struct
...
        let cell = E.overwrite_value old e.atom e.v in
        let co_cell = Array.copy st.co_cell in
-       let cell2 = match n.edge.E.edge with
+       let cell2 = match n.prev.edge.E.edge with
         | E.Rmw rmw -> E.compute_rmw rmw old cell
-        | _ -> if rv = -1 then cell else rv in
-      co_cell.(0) <- cell2;
+        | _ -> cell in
+       co_cell.(0) <- cell2;
       {e with cell=co_cell;},{ st with co_cell; }
```
Comment (maranget) [inline gen/cycle.ml]: Performing a RWM operation is as follows:

1. Read old value, here `let old = st.co_cell.(0)`
2. Compute, here `let cell2 = ... E.compute_rmw old e.v`
3. Store the result, here  `co_cell .(0) <- E.overwrite_value old e.atom cell2`

The `E.overwrite` function replaces the the subpart of the memory location specified by `e.atom` with the operation result, leaving the rest unchanged. This applies only to mixed-size mode, where `e.atom` may specify some subpart such as second byte, first 16bits word, etc.

File: gen/cycle.ml [hunk 13]
```diff
@@ -463,16 +463,16 @@ module CoSt = struct
...
 
-  let set_cell ?(rv = -1) st n =
+  let set_cell st n =
     let e = n.evt in match e.bank with
     | Ord -> begin
        let old = st.co_cell.(0) in
        let cell = E.overwrite_value old e.atom e.v in
        let co_cell = Array.copy st.co_cell in
-       let cell2 = match n.edge.E.edge with
+       let cell2 = match n.prev.edge.E.edge with
```
Comment (maranget) [inline gen/cycle.ml:472]: This ` .prev` is surprising.

File: gen/cycle.ml [hunk 14]
```diff
@@ -787,27 +787,9 @@ let set_same_loc st n0 =
...
-                   let e,st1 = CoSt.set_cell st1 n in
-                   n.evt <- e ;
-                   let co = n.evt.cell.(0) in
-                   do_set_write_val st1 pte_val ns ~rv:co
-                | _   -> do_set_write_val st pte_val ns
-              end
-            | Code _ -> do_set_write_val st pte_val ns
-          end
-        | Some J |None -> do_set_write_val st pte_val ns
+        | Some R | Some J |None -> do_set_write_val st pte_val ns
```
Comment (maranget) [inline gen/cycle.ml]: A matter of style...
```suggestion
        | Some (R|J) |None -> do_set_write_val st pte_val ns
```

File: gen/AArch64Arch_gen.ml [hunk 15]
```diff
@@ -604,7 +604,9 @@ let compute_rmw r old co = match r with
         | A_ADD -> old + co
         | A_SMAX -> if old > co then old else co
         | A_SMIN -> if old < co then old else co
-        | A_EOR | A_SET | A_CLR -> co
+        | A_EOR -> old lxor co
+        | A_SET -> old lor co
+        | A_CLR -> lnot (old land co)
```
Comment (maranget) [inline gen/AArch64Arch_gen.ml:609]: Isn't it `old land (lnot co)`?

Comment (hugookeeffe) [reply gen/AArch64Arch_gen.ml:609]: Ah yes, that is my bad. I will change that.

File: gen/cycle.ml (hunk 12 above)
Comment (hugookeeffe) [reply gen/cycle.ml]: Thank you!

File: gen/cycle.ml [hunk 16]
```diff
@@ -463,13 +463,18 @@ module CoSt = struct
...
+    let e = n.evt in match e.bank with
+    | Ord -> begin
        let old = st.co_cell.(0) in
        let cell = E.overwrite_value old e.atom e.v in
        let co_cell = Array.copy st.co_cell in
-       co_cell.(0) <- cell ;
-       {e with cell=co_cell;},{ st with co_cell; }
+       let cell2 = match n.prev.edge.E.edge with
+        | E.Rmw rmw -> E.compute_rmw rmw old e.v
+        | _ -> cell in
```
Comment (relokin) [inline gen/cycle.ml:474]: Shouldn't this be:
```
| _ -> e.v
```
In which case we can also get rid of line 470?

**Review by relokin — COMMENTED** [review]
Thanks for all the changes Hugo. I have one comment but otherwise, it looks good to me and runs fine with a couple of examples I tried.

@maranget how do you typically merge? Should Hugo squash all commits into one and then we can use fast-forward merge? Or will the merge squash them automatically.

File: gen/cycle.ml (hunk 16 above)
Comment (maranget) [reply gen/cycle.ml:474]: Hi @relokin, I usually merge with a merge commit. If commit history is complex I squash. I mean I use github "squash and merge" button. Here I do not know yet.

Comment (maranget) [reply gen/cycle.ml:474]: Hi @hugookeeffe and @relokin, I'd like to merge the PR in once the last comment by @relokin above is handled. 

Notice that, once simplified, the `set_cell` function can be adapted easily to mixed-size by using the `E.extract_value` function. For clarity, commit ba305dfc1eb3ffd57087c04e57b6def2c65eee6a performs the adaptation (also addressing @relokin's comment).

Comment (hugookeeffe) [reply gen/cycle.ml]: Hi @maranget, I have made the changes for the last comment in my recent commit. Thank you for the help!

Comment (maranget) [reply gen/cycle.ml]: Thanks, I'll merge tomorrow.

**Comment by maranget** [comment]
Merge after squash, thanks.

**Pull request merged**